packages/theme/src/cli/services/update/run.ts

```diff
@@ -0,0 +1,178 @@
+  const ctx = await renderTasks<Context>([
+    task(`Checking your theme update script`, (ctx) => check(ctx, options)),
+    task(`Starting your theme update`, (ctx) => triggerUpdater(ctx, session, options)),
+    task(`Updating your theme`, (ctx) => waitForUpdater(ctx, session)),
+  ])
+
+  renderSuccess({
+    body: 'Your theme has been updated.',
+    nextSteps: [
+      [
+        {
+          link: {
+            label: 'Explore the updated theme in the code editor',
+            url: codeEditorUrl(ctx.theme, session),
+          },
+        },
+      ],
+      [
+        {
+          link: {
+            label: 'Explore the updated theme in the theme editor',
+            url: themeEditorUrl(ctx.theme, session),
+          },
+        },
+      ],
+      [
+        {
+          link: {
+            label: 'Preview the updated theme',
+            url: themePreviewUrl(ctx.theme, session),
+          },
+        },
+      ],
+    ],
+  })
+}
+
+async function check(ctx: Context, options: Options) {
+  const {script} = options
+
+  await sleep(1)
+  await checkScript(script)
+
+  ctx.scriptContent = await readFile(joinPath(script))
+}
+
+async function triggerUpdater(ctx: Context, session: AdminSession, options: Options) {
+  const {'source-theme': source, 'target-theme': target} = options
+  const store = session.storeFqdn
+
+  const themes = await fetchStoreThemes(session)
+
+  const sourceThemeId = findThemeByIdentifier(store, themes, source).id
+  const targetThemeId = findThemeByIdentifier(store, themes, target).id
+
+  // This is a false positive, as `renderTasks` performs tasks sequentially.
+  // eslint-disable-next-line require-atomic-updates
+  ctx.theme = await triggerUpdaterAPI(session, ctx.scriptContent, sourceThemeId, targetThemeId)
+}
+
+async function waitForUpdater(ctx: Context, session: AdminSession) {
+  const startTime = Date.now()
+
+  /**
+   * Generally, it's a good practice to take full advantage of the parallelization
+   * benefits of async/await.
+   *
+   * However, in the context of polling an API, each iteration must wait for the
+   * response, before the next polling.
+   */
+  // eslint-disable-next-line no-await-in-loop
+  while (await isUpdaterIsProgress(ctx, session, startTime)) {
+    // retry
+  }
+}
+
+async function isUpdaterIsProgress(ctx: Context, session: AdminSession, startTime: number) {
+  const theme = await fetchTheme(ctx.theme.id, session)
+
+  if (!theme) {
+    throw new AbortError(
+      'The `update_extension.json` script could not be executed due to a runtime issue.',
+      'Please check the `update_extension.json` script and retry.',
+    )
+  }
+
+  const elapsedTime = Date.now() - startTime
+
+  if (theme.processing && elapsedTime >= UPDATER_TIMEOUT) {
+    throw new AbortError(
+      'The `update_extension.json` script could not be executed due to a timeout issue.',
+      'Please check the `update_extension.json` script and retry.',
```

Comment:
same comment as above

Reply:
This is being handled [here](https://github.com/Shopify/cli/pull/1946#discussion_r1217700057).